// _@autorest_typescript@6.0.0-beta.7@@autorest/typescript/dist/src/generators/static/apiExtractorConfig.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateApiExtractorConfig = void 0;
const autorestSession_1 = require("../../autorestSession");
function generateApiExtractorConfig(project) {
    const { generateMetadata } = autorestSession_1.getAutorestOptions();
    if (!generateMetadata) {
        return;
    }
    const config = {
        $schema: "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
        mainEntryPointFilePath: `./esm/index.d.ts`,
        docModel: {
            enabled: true
        },
        apiReport: {
            enabled: true,
            reportFolder: "./review"
        },
        dtsRollup: {
            enabled: true,
            untrimmedFilePath: "",
            publicTrimmedFilePath: `./esm/index.d.ts`
        },
        messages: {
            tsdocMessageReporting: {
                default: {
                    logLevel: "none"
                }
            },
            extractorMessageReporting: {
                "ae-missing-release-tag": {
                    logLevel: "none"
                },
                "ae-unresolved-link": {
                    logLevel: "none"
                }
            }
        }
    };
    project.createSourceFile("api-extractor.json", JSON.stringify(config), {
        overwrite: true
    });
}
exports.generateApiExtractorConfig = generateApiExtractorConfig;
