// _@autorest_typescript@6.0.0-beta.7@@autorest/typescript/dist/src/transforms/groupTransforms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformGroups = void 0;
const codemodel_1 = require("@autorest/codemodel");
const languageHelpers_1 = require("../utils/languageHelpers");
const modelDetails_1 = require("../models/modelDetails");
const objectTransforms_1 = require("./objectTransforms");
function transformGroups(codeModel) {
    var _a;
    codeModel.schemas.groups = codeModel.schemas.groups || [];
    return ((_a = codeModel.schemas.groups) !== null && _a !== void 0 ? _a : []).map(group => transformGroup(group));
}
exports.transformGroups = transformGroups;
function transformGroup({ language, properties }) {
    const { name, description } = languageHelpers_1.getLanguageMetadata(language);
    const groupProperties = (properties || []).map(prop => objectTransforms_1.transformProperty(prop));
    return {
        children: [],
        parents: [],
        hasAdditionalProperties: false,
        kind: modelDetails_1.ObjectKind.Basic,
        name,
        // Groups are created by modelerfour so they don't have a serializedName
        serializedName: name,
        description: description || undefined,
        schema: new codemodel_1.ObjectSchema(name, description),
        properties: groupProperties
    };
}
