// _@autorest_typescript@6.0.0-beta.7@@autorest/typescript/dist/src/generators/clientFileGenerator.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkForNameCollisions = exports.generateClient = void 0;
const operationGenerator_1 = require("./operationGenerator");
const nameUtils_1 = require("../utils/nameUtils");
const codemodel_1 = require("@autorest/codemodel");
const parameterUtils_1 = require("./utils/parameterUtils");
const importUtils_1 = require("./utils/importUtils");
const responseTypeUtils_1 = require("./utils/responseTypeUtils");
const tracingUtils_1 = require("./utils/tracingUtils");
const pagingOperations_1 = require("./utils/pagingOperations");
const autorestSession_1 = require("../autorestSession");
function generateClient(clientDetails, project) {
    const { useCoreV2, hideClients, srcPath, addCredentials } = autorestSession_1.getAutorestOptions();
    const clientContextClassName = `${clientDetails.className}Context`;
    const hasMappers = !!clientDetails.mappers.length;
    // Check if there are any client level operations
    const inlineOperations = clientDetails.operationGroups.filter(og => og.isTopLevel);
    const hasInlineOperations = inlineOperations.length > 0;
    // Check if there are any non client-level operations to import
    const importedOperations = clientDetails.operationGroups.filter(og => !og.isTopLevel);
    const hasImportedOperations = importedOperations.length > 0;
    if (hasImportedOperations && hasInlineOperations) {
        // Check if there is a name collision between client-level operation names
        // and operation group key names.
        checkForNameCollisions(importedOperations, inlineOperations);
    }
    const hasCredentials = !!addCredentials;
    const hasClientOptionalParams = clientDetails.parameters.some(p => !p.required && p.implementationLocation === codemodel_1.ImplementationLocation.Client);
    const clientFile = project.createSourceFile(`${srcPath}/${clientDetails.sourceFileName}.ts`, undefined, {
        overwrite: true
    });
    const flattenedInlineOperations = inlineOperations.reduce((acc, curr) => (acc = [...acc, ...curr.operations]), []);
    if (hasCredentials || hasInlineOperations || !hasClientOptionalParams) {
        if (!useCoreV2) {
            clientFile.addImportDeclaration({
                namespaceImport: "coreHttp",
                moduleSpecifier: "@azure/core-http"
            });
        }
        else {
            clientFile.addImportDeclaration({
                namespaceImport: "coreClient",
                moduleSpecifier: "@azure/core-client"
            });
            clientFile.addImportDeclaration({
                namespaceImport: "coreRestPipeline",
                moduleSpecifier: "@azure/core-rest-pipeline"
            });
            clientFile.addImportDeclaration({
                namespaceImport: "coreTracing",
                moduleSpecifier: "@azure/core-tracing"
            });
            if (hasCredentials) {
                clientFile.addImportDeclaration({
                    namespaceImport: "coreAuth",
                    moduleSpecifier: "@azure/core-auth"
                });
            }
        }
    }
    pagingOperations_1.addPagingEsNextRef(flattenedInlineOperations, clientFile);
    pagingOperations_1.addPagingImports(flattenedInlineOperations, clientFile);
    const hasLro = inlineOperations.some(og => og.operations.some(o => o.isLro));
    if (hasInlineOperations && hasLro) {
        clientFile.addImportDeclaration({
            namedImports: ["PollerLike", "PollOperationState"],
            moduleSpecifier: "@azure/core-lro"
        });
        clientFile.addImportDeclaration({
            namedImports: ["LroEngine"],
            moduleSpecifier: `./lro`
        });
        if (useCoreV2) {
            clientFile.addImportDeclaration({
                namedImports: ["CoreClientLro", "shouldDeserializeLro"],
                moduleSpecifier: `./coreClientLro`
            });
        }
        else {
            clientFile.addImportDeclaration({
                namedImports: ["CoreHttpLro", "shouldDeserializeLro"],
                moduleSpecifier: `./coreHttpLro`
            });
        }
    }
    if (hasImportedOperations) {
        clientFile.addImportDeclaration({
            namedImports: importedOperations.map(o => `${nameUtils_1.normalizeName(o.name, nameUtils_1.NameType.OperationGroup, true /* shouldGuard */)}Impl`),
            moduleSpecifier: "./operations"
        });
        clientFile.addImportDeclaration({
            namedImports: importedOperations.map(o => `${nameUtils_1.normalizeName(o.name, nameUtils_1.NameType.OperationGroup, true /* shouldGuard */)}`),
            moduleSpecifier: "./operationsInterfaces"
        });
    }
    if (hasInlineOperations && importUtils_1.shouldImportParameters(clientDetails)) {
        tracingUtils_1.addTracingOperationImports(clientFile, ".");
        clientFile.addImportDeclaration({
            namespaceImport: "Parameters",
            moduleSpecifier: "./models/parameters"
        });
    }
    // Only import mappers if there are any
    if (hasInlineOperations && hasMappers) {
        clientFile.addImportDeclaration({
            namespaceImport: "Mappers",
            moduleSpecifier: "./models/mappers"
        });
    }
    clientFile.addImportDeclaration({
        namedImports: [clientContextClassName],
        moduleSpecifier: `./${clientDetails.sourceFileName}Context`
    });
    const clientClass = clientFile.addClass({
        name: clientDetails.className,
        extends: clientContextClassName,
        isExported: true
    });
    if (hideClients) {
        clientClass.addJsDoc({
            tags: [
                {
                    tagName: "internal"
                }
            ]
        });
    }
    const importedModels = new Set();
    writeConstructor(clientDetails, clientClass, importedModels);
    writeClientOperations(clientFile, clientClass, clientDetails, hasLro, importedModels);
    // Use named import from Models
    if (importedModels.size) {
        clientFile.addImportDeclaration({
            namedImports: [...importedModels],
            moduleSpecifier: "./models"
        });
    }
    clientFile.fixUnusedIdentifiers();
}
exports.generateClient = generateClient;
function checkForNameCollisions(importedOperations, inlineOperations) {
    const groupOpsKeyNames = importedOperations.map(og => og.key.toLowerCase());
    const inlineOpsKeyNames = inlineOperations.map(og => og.operations.map(operation => operation.name.toLowerCase()));
    const collidingKeyNames = inlineOpsKeyNames.map(inlineOpsKeyArray => inlineOpsKeyArray.filter(inlineOpKey => groupOpsKeyNames.includes(inlineOpKey)));
    if (collidingKeyNames.length > 0 && collidingKeyNames[0].length > 0) {
        const messages = collidingKeyNames.map(key => `Operation Group(s) '${key}' is/are colliding with client-level operation(s) with the same name.`);
        throw new Error(...messages);
    }
}
exports.checkForNameCollisions = checkForNameCollisions;
function writeConstructor(clientDetails, classDeclaration, importedModels) {
    const requiredParams = clientDetails.parameters.filter(param => param.required &&
        param.implementationLocation === codemodel_1.ImplementationLocation.Client &&
        !param.defaultValue &&
        param.schemaType !== codemodel_1.SchemaType.Constant);
    const hasClientOptionalParameters = clientDetails.parameters.some(param => !param.required &&
        param.implementationLocation === codemodel_1.ImplementationLocation.Client);
    const docs = [
        `Initializes a new instance of the ${clientDetails.className} class.`,
        ...parameterUtils_1.formatJsDocParam(requiredParams),
        `@param options The parameter options`
    ]
        .filter(d => !!d)
        .join("\n");
    let optionsParameterType = "ServiceClientOptions";
    if (hasClientOptionalParameters) {
        const paramType = `${clientDetails.className}OptionalParams`;
        importedModels.add(paramType);
        optionsParameterType = paramType;
    }
    requiredParams.forEach(({ typeDetails }) => typeDetails.usedModels.forEach(model => importedModels.add(model)));
    const clientConstructor = classDeclaration.addConstructor({
        docs: [docs],
        parameters: [
            ...requiredParams.map(p => ({
                name: p.name,
                hasQuestionToken: !p.required,
                type: p.typeDetails.typeName
            })),
            {
                name: "options",
                hasQuestionToken: true,
                type: optionsParameterType
            }
        ]
    });
    clientConstructor.addStatements([
        `super(${[...requiredParams.map(p => p.name), "options"].join()});`
    ]);
    const operationDeclarationDetails = getOperationGroupsDeclarationDetails(clientDetails.operationGroups.filter(og => !og.isTopLevel));
    clientConstructor.addStatements([
        ...operationDeclarationDetails.map(({ name, typeName }) => `this.${name} = new ${typeName}Impl(this)`)
    ]);
}
function getOperationGroupsDeclarationDetails(operationGroups) {
    return operationGroups.map(og => {
        return {
            name: nameUtils_1.normalizeName(og.name, nameUtils_1.NameType.Property),
            typeName: `${nameUtils_1.normalizeName(og.name, nameUtils_1.NameType.OperationGroup, true /* shouldGuard */)}`
        };
    });
}
function writeClientOperations(file, classDeclaration, clientDetails, hasLro, importedModels) {
    const allModelsNames = responseTypeUtils_1.getAllModelsNames(clientDetails);
    const topLevelGroup = clientDetails.operationGroups.find(og => og.isTopLevel);
    const hasMappers = !!clientDetails.mappers.length;
    // Add top level operation groups as client properties
    if (!!topLevelGroup) {
        if (hasLro) {
            operationGenerator_1.writeGetOperationOptions(classDeclaration);
        }
        operationGenerator_1.writeOperations(topLevelGroup, classDeclaration, importedModels, allModelsNames, clientDetails, true // isInline
        );
        operationGenerator_1.addOperationSpecs(topLevelGroup, file, clientDetails.parameters, hasMappers);
    }
    const operationsDeclarationDetails = getOperationGroupsDeclarationDetails(clientDetails.operationGroups.filter(og => !og.isTopLevel));
    // Each operation group will have its class
    // and the client class will have each group as properties
    classDeclaration.addProperties(operationsDeclarationDetails.map(op => {
        return {
            name: op.name,
            type: op.typeName
        };
    }));
}
