// _@autorest_codemodel@4.15.0@@autorest/codemodel/dist/model/common/code-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeModel = void 0;
const metadata_1 = require("./metadata");
const schemas_1 = require("./schemas");
const info_1 = require("./info");
const operation_1 = require("./operation");
const codegen_1 = require("@azure-tools/codegen");
const linq_1 = require("@azure-tools/linq");
const security_1 = require("./security");
class CodeModel extends metadata_1.Metadata {
    constructor(title, sourceTracking = false, objectInitializer) {
        super();
        // if we are enabling source tracking, then we have to use a proxied version of this
        const $this = sourceTracking ? codegen_1.enableSourceTracking(this) : this;
        $this.info = new info_1.Info(title);
        $this.schemas = new schemas_1.Schemas();
        $this.operationGroups = [];
        $this.security = new security_1.Security(false);
        this.applyTo($this, objectInitializer);
    }
    get globals() {
        return this.globalParameters || (this.globalParameters = []);
    }
    getOperationGroup(group) {
        let result = this.operationGroups.find((each) => group.toLowerCase() === each.$key.toLowerCase());
        if (!result) {
            result = new operation_1.OperationGroup(group);
            this.operationGroups.push(result);
        }
        return result;
    }
    findGlobalParameter(predicate) {
        return this.globals.find(predicate);
    }
    addGlobalParameter(predicateOrParameter, create = undefined) {
        try {
            if (typeof predicateOrParameter !== "function") {
                // overload : parameter passed
                this.globals.push(predicateOrParameter);
                return predicateOrParameter;
            }
            // overload : predicate, parameter passed
            let p = this.findGlobalParameter(predicateOrParameter);
            if (!p) {
                this.globals.push((p = linq_1.realize(create)));
            }
            return p;
        }
        finally {
            this.globalParameters = linq_1.sort.numericly.ascendingInvalidLast(this.globals, (each) => { var _a; return (_a = each.extensions) === null || _a === void 0 ? void 0 : _a["x-ms-priority"]; });
        }
    }
}
exports.CodeModel = CodeModel;
